File: 2026/05-May/28-Thursday/solution.go
type TrieNode struct {
	children map[byte]*TrieNode
	idx      int
	length   int
}

func newTrieNode() *TrieNode {
	return &TrieNode{
		children: make(map[byte]*TrieNode),
		idx:      -1,
		length:   int(^uint(0) >> 1),
	}
}

func update(node *TrieNode, idx int, length int) {
	if length < node.length {
		node.length = length
		node.idx = idx
	}
}

func stringIndices(wordsContainer []string, wordsQuery []string) []int {

	root := newTrieNode()

	for i, word := range wordsContainer {

		node := root
		update(node, i, len(word))

		for j := len(word) - 1; j >= 0; j-- {
			ch := word[j]

			if _, ok := node.children[ch]; !ok {
				node.children[ch] = newTrieNode()
			}

			node = node.children[ch]
			update(node, i, len(word))
		}
	}

	ans := []int{}

	for _, word := range wordsQuery {

		node := root

		for j := len(word) - 1; j >= 0; j-- {
			ch := word[j]

			if _, ok := node.children[ch]; !ok {
				break
			}

			node = node.children[ch]
		}

		ans = append(ans, node.idx)
	}

	return ans
}

File: 2026/05-May/28-Thursday/solution.js
class TrieNode {
    constructor() {
        this.children = {};
        this.idx = -1;
        this.length = Infinity;
    }
}

var stringIndices = function(wordsContainer, wordsQuery) {

    const root = new TrieNode();

    function update(node, idx, len) {
        if (len < node.length) {
            node.length = len;
            node.idx = idx;
        }
    }

    for (let i = 0; i < wordsContainer.length; i++) {
        let word = wordsContainer[i];

        let node = root;
        update(node, i, word.length);

        for (let j = word.length - 1; j >= 0; j--) {
            let ch = word[j];

            if (!node.children[ch]) {
                node.children[ch] = new TrieNode();
            }

            node = node.children[ch];
            update(node, i, word.length);
        }
    }

    let ans = [];

    for (let word of wordsQuery) {
        let node = root;

        for (let j = word.length - 1; j >= 0; j--) {
            let ch = word[j];

            if (!node.children[ch]) break;

            node = node.children[ch];
        }

        ans.push(node.idx);
    }

    return ans;
};
